server/src/main/java/org/apache/lucene/index/BaseSingleStarTreeBuilder.java

```diff
@@ -0,0 +1,660 @@
+import org.apache.lucene.codecs.DocValuesProducer;
+import org.apache.lucene.search.DocIdSetIterator;
+import org.opensearch.index.compositeindex.datacube.DateDimension;
+import org.opensearch.index.compositeindex.datacube.Dimension;
+import org.opensearch.index.compositeindex.datacube.Metric;
+import org.opensearch.index.compositeindex.datacube.MetricStat;
+import org.opensearch.index.compositeindex.datacube.startree.StarTreeField;
+import org.opensearch.index.compositeindex.datacube.startree.StarTreeFieldConfiguration;
+import org.opensearch.index.compositeindex.datacube.startree.aggregators.MetricStatFieldPair;
+import org.opensearch.index.compositeindex.datacube.startree.aggregators.ValueAggregator;
+import org.opensearch.index.compositeindex.datacube.startree.aggregators.ValueAggregatorFactory;
+import org.opensearch.index.compositeindex.datacube.startree.aggregators.numerictype.StarTreeNumericType;
+import org.opensearch.index.compositeindex.datacube.startree.builder.SingleTreeBuilder;
+import org.opensearch.index.compositeindex.datacube.startree.builder.StarTreeDocValuesIteratorFactory;
+import org.opensearch.index.compositeindex.datacube.startree.data.StarTreeDocument;
+import org.opensearch.index.compositeindex.datacube.startree.utils.StarTreeBuilderUtils;
+import org.opensearch.index.fielddata.IndexNumericFieldData;
+import org.opensearch.index.mapper.Mapper;
+import org.opensearch.index.mapper.MapperService;
+import org.opensearch.index.mapper.NumberFieldMapper;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * Base class for star-tree builder
+ */
+public abstract class BaseSingleStarTreeBuilder implements SingleTreeBuilder {
+
+    // TODO: STAR_TREE_CODEC will be moved to CodecService once the Star Tree Codec is defined
+    public static final String STAR_TREE_CODEC = "startreecodec";
+
+    private static final Logger logger = LogManager.getLogger(BaseSingleStarTreeBuilder.class);
+
+    public static final int STAR_IN_DOC_VALUES_INDEX = -1;
+
+    protected final String[] dimensionsSplitOrder;
+    protected final Set<Integer> skipStarNodeCreationForDimensions;
+    protected final String[] metrics;
+
+    protected final int numMetrics;
+    protected final int numDimensions;
+    protected int numDocs;
+    protected int totalDocs;
+    protected int numNodes;
+    protected final int maxLeafDocuments;
+
+    protected final StarTreeBuilderUtils.TreeNode rootNode = getNewNode();
+
+    // TODO: This will be initialized with OnHeap / OffHeap Implementations (Commented it's occurrences for now)
+    // private IndexOutput indexOutput;
+    protected DocIdSetIterator[] dimensionReaders;
+    protected DocIdSetIterator[] metricReaders;
+
+    protected ValueAggregator[] valueAggregators;
```

Comment:
Lets form a complex object that contains 
```
metric field name
metric
value aggregator
numeric type
```

And then we can iterate over the same ?